### PR TITLE
test: add consumer e2e test for manager-api

### PR DESCRIPTION
**Pull request by idbeta**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] add consumer e2e test for manager-api

- Depend on issues #568 #870 #852 
- Related issues
https://github.com/apache/apisix-dashboard/issues/607

### CODE REVIEW

File: api/test/e2e/consumer_test.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+			ExpectStatus: http.StatusOK,
+		},
+		{
+			caseDesc:     "verify route",
+			Object:       APISIXExpect(t),
+			Method:       http.MethodGet,
+			Path:         "/hello",
+			Headers:      map[string]string{"apikey": "auth-one"},
+			ExpectStatus: http.StatusOK,
+			Sleep:        sleepTime, //sleep x millisecond before verify route
```
Comment (nic-chen) [inline api/test/e2e/consumer_test.go:71]: need to check body.

Comment (nic-chen) [reply api/test/e2e/consumer_test.go:71]: need to check before consumer created.

File: api/test/e2e/consumer_test.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+		},
+	}
+
+	for _, tc := range tests {
+		testCaseCheck(tc)
+	}
+}
+
+//CASE 3: add consumer with labels
+
```
Comment (nic-chen) [inline api/test/e2e/consumer_test.go:124]: why leave this empty ?

File: api/test/e2e/consumer_test.go [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+func TestConsumer_delete_consumer(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc:     "delete consumer",
+			Object:       MangerApiExpect(t),
+			Method:       http.MethodDelete,
+			Path:         "/apisix/admin/consumers/jack",
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+		},
```
Comment (nic-chen) [inline api/test/e2e/consumer_test.go:135]: need a test case for deleting an un-exist consumer.

**Review by nic-chen — CHANGES_REQUESTED** [review]

File: api/test/e2e/consumer_test.go (hunk 3 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go:146]: ok, thanks for the suggestion.

File: api/test/e2e/consumer_test.go (hunk 2 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go]: already added.

File: api/test/e2e/consumer_test.go (hunk 1 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go]: ok, thanks for the suggestion, I already add it now.

File: api/test/e2e/consumer_unusual_test.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+//case 8: update consumer, check if updatetime is updated
+func TestConsumer_create_consumer_with_createtime_updatetime(t *testing.T) {
+	//create consumer, save the result_A ( create_time and update_time )
+	data := `{
+		"username":"case_8",
+		"desc": "new consumer"
+   }`
+	request, _ := http.NewRequest("PUT", "http://127.0.0.1:8080/apisix/admin/consumers", strings.NewReader(data))
```
Comment (LiteSun) [inline api/test/e2e/consumer_unusual_test.go]: Is it better to extract public variables? such as `BASE_URL = http://127.0.0.1:8080/apisix/admin/consumers`

File: api/test/e2e/consumer_test.go [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/tidwall/gjson"
+)
+
+//case 1: add consumer with username
```
Comment (membphis) [inline api/test/e2e/consumer_test.go]: remove useless code

Comment (membphis) [reply api/test/e2e/consumer_test.go]: and pls remove other similar points

File: api/test/e2e/consumer_test.go [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+						 "key": "auth-one"
+					 }
+				 },
+				 "desc": "test description"
+			 }`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusOK,
+		},
+		{
+			caseDesc:     "verify route",
```
Comment (membphis) [inline api/test/e2e/consumer_test.go:79]: desc: hit route with correct `key`

File: api/test/e2e/consumer_test.go [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+			Object:       APISIXExpect(t),
+			Method:       http.MethodGet,
+			Path:         "/hello",
+			Headers:      map[string]string{"apikey": "auth-one"},
+			ExpectStatus: http.StatusOK,
+			ExpectBody:   "hello world",
+			Sleep:        sleepTime,
+		},
+		{
+			caseDesc:     "verify route",
```
Comment (membphis) [inline api/test/e2e/consumer_test.go:89]: desc: hit route with incorrect `key`

File: api/test/e2e/consumer_test.go [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+			Object:       APISIXExpect(t),
+			Method:       http.MethodGet,
+			Path:         "/hello",
+			Headers:      map[string]string{"apikey": "auth-new"},
+			ExpectStatus: http.StatusUnauthorized,
+			ExpectBody:   "Invalid API key in request",
+			Sleep:        sleepTime,
+		},
```
Comment (membphis) [inline api/test/e2e/consumer_test.go:97]: need to delete the `consumer` and check it in dataplane

File: api/test/e2e/consumer_test.go [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+	for _, tc := range tests {
+		testCaseCheck(tc)
+	}
+}
+
+//case 2: add consumer without username
+func TestConsumer_add_consumer_without_username(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc: "create consumer",
```
Comment (membphis) [inline api/test/e2e/consumer_test.go:108]: bad title.

`create consumer without user name`

File: api/test/e2e/consumer_test.go [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+						 "key": "auth-new"
+					 }
+				 },
+				 "desc": "test description"
+			 }`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusBadRequest,
+		},
+		{
+			caseDesc:     "verify route",
```
Comment (membphis) [inline api/test/e2e/consumer_test.go:124]: I think we need to create the `route` first

**Review by membphis — COMMENTED** [review]
we can close this PR, split this big PR to smaller ones

**Pull request closed without merging**